docs/httpbench.1
.\" Automatically generated by Pod::Man 2.25 (Pod::Simple 3.16)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.ie \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.el \{\
.    de IX
..
.\}
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "HTTPBENCH 1"
.TH HTTPBENCH 1 "2014-06-22" "httpbench 0.2.2" "User Commands"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
httpbench \- Simple HTTP benchmarking tool
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
httpbench \-u <url or urllist.txt> \-d sec \-c concurrent \-r rps [\-t ms] [\-e expected]
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\s-1HTTP\s0 Bench is a small program for Linux, \s-1UNIX\s0 and a like written in C which can be used to benchmark \s-1HTTP\s0 servers. For more infos please consult the project website.
.SS "Arguments"
.IX Subsection "Arguments"
\fIMandatory\fR
.IX Subsection "Mandatory"
.ie n .IP """\-u url or urllist.txt""" 12
.el .IP "\f(CW\-u url or urllist.txt\fR" 12
.IX Item "-u url or urllist.txt"
An \s-1URL\s0 to benchmark if argument begins with http:// or https:// or a path to a text file containing a list of all HTTP-URLs to request during a benchmark. Each line should start with http:// or https://.
.ie n .IP """\-d seconds""" 12
.el .IP "\f(CW\-d seconds\fR" 12
.IX Item "-d seconds"
Duration of the benchmark in seconds. E.g. 300 for a five minutes benchmark.
.ie n .IP """\-c concurrent""" 12
.el .IP "\f(CW\-c concurrent\fR" 12
.IX Item "-c concurrent"
Number of cuncurrent threads used by the benchmark.
.ie n .IP """\-r rps""" 12
.el .IP "\f(CW\-r rps\fR" 12
.IX Item "-r rps"
Requests per second wanted for the benchmark.
.PP
\fIOptional\fR
.IX Subsection "Optional"
.ie n .IP """\-t timeoutms""" 12
.el .IP "\f(CW\-t timeoutms\fR" 12
.IX Item "-t timeoutms"
Timeout in milliseconds for each request. This timeout will actually not interrupt any requests. It's there just for statistical reasons. E.g. how many requests take longer than x milliseconds?
.ie n .IP """\-e expectedresponsestr""" 12
.el .IP "\f(CW\-e expectedresponsestr\fR" 12
.IX Item "-e expectedresponsestr"
The expected response of a request. E.g. all requests should contain \s-1HTML\s0 in its reponse. All responses not matching are counted and reported.
.SS "\s-1EXAMPLES\s0"
.IX Subsection "EXAMPLES"
.ie n .IP """httpbench \-u mylist.txt \-d 60 \-c 10 \-r 1000""" 12
.el .IP "\f(CWhttpbench \-u mylist.txt \-d 60 \-c 10 \-r 1000\fR" 12
.IX Item "httpbench -u mylist.txt -d 60 -c 10 -r 1000"
Iterates over each url in mylist.txt with 10 cuncurrent threads for 60 seconds and tries to reach 1000 requests per second. Each curl failure will be reported at the end of the test (total count and total %).
.ie n .IP """httpbench \-u http://www.example.com \-d 300 \-c 1000 \-r 10000 \-t 20 \-e ""Hello World""""" 12
.el .IP "\f(CWhttpbench \-u http://www.example.com \-d 300 \-c 1000 \-r 10000 \-t 20 \-e ``Hello World''\fR" 12
.IX Item "httpbench -u http://www.example.com -d 300 -c 1000 -r 10000 -t 20 -e ""Hello World"""
Benchmarks the single \s-1URL\s0 http://www.example.com with 1000 cuncurrent threads for 5 minutes and tries to reach 10000 requests per second. Each request which takes longer than 20 milliseconds will be reported at the end of the test (total count and total %) and each request which does not contain \*(L"Hello World\*(R" in its response string will also be reported (total count and total %). The curl errors will also be reported.
.SH "LICENSE"
.IX Header "LICENSE"
See package description or project website.
.SH "AUTHOR"
.IX Header "AUTHOR"
Paul Buetow \- <http://httpbench.buetow.foo>


CHANGELOG
Fri Apr  6 11:14:04 CEST 2012
* Initial git checkin, no release yet


src/httpbench.c
// Small humble program for benchmarking
// (C) 2012 Dipl.-Inform. (FH) Paul C. Buetow
// Look in ./debian/copyright for license info

#include "httpbench.h"

void synopsis(void) {
  printf("httpbench %s synopsis:\n", VERSION);
  printf("httpbench -u <url or urllist.txt> -d sec -c concurrent -r rps [-t ms] [-e expected]\n"); 
  printf("Please also consult the httpbench manual page\n");
}

void checkarg_c(char c_name, char *c_arg) {
  if (c_arg == NULL) {
    fprintf(stderr, "Missing mandatory value for option '%c'\n", c_name);
    synopsis();
    exit(E_MISSING_OPT_C);
  }
}

void checkarg_i(char c_name, int i_arg) {
  if (i_arg == -1) {
    fprintf(stderr, "Missing mandatory value for option '%c'\n", c_name);
    synopsis();
    exit(E_MISSING_OPT_I);
  }
}

void print_stats(
    unsigned int ui_count,
    int i_elapsed_time,
    double d_time_max,
    double d_time_min,
    double d_time_avg,
    unsigned int ui_timeout_exceeded,
    unsigned int ui_curl_errors,
    unsigned int ui_parse_errors) {

  fprintf(stdout, "(thread %d) count:%d elapsed:%d (%.1f rps), max:%0.6f min:%0.6f avg:%0.6f timeout_exceeded:%d (%.3f%%) curl_errors:%d (%.3f%%) parse_errors:%d (%.3f%%)\n",
      (int) pthread_self(),
      ui_count,
      i_elapsed_time,
      ((double) ui_count/(double) i_elapsed_time),
      d_time_max,
      d_time_min,
      d_time_avg / (double) ui_count,
      ui_timeout_exceeded,
      ((double) ui_timeout_exceeded)/(((double) ui_count)/100),
      ui_curl_errors,
      ((double) ui_curl_errors)/(((double) ui_count)/100),
      ui_parse_errors,
      ((double) ui_parse_errors)/(((double) ui_count)/100)
      );
}

// Very *very* magic algorithm!
double sleep_us(double d_sleep_us, double d_rps, double d_rps_wanted, double *pd_perc) {
  *pd_perc = d_rps / d_rps_wanted;

  if (*pd_perc < 0.1)
    return d_sleep_us * 0.1;
  else if (*pd_perc > 2.0)
    return d_sleep_us * 2;

  return *pd_perc * d_sleep_us;
}

size_t static write_data(void *p_buf, size_t s_size, size_t s_nmemb, void *p_val) {
  char **pc_response = (char**) p_val;

  *pc_response = strndup(p_buf, (size_t)(s_size * s_nmemb));

  return s_size * s_nmemb;
}

CURL* curl_init(char **c_response) {
  CURL *p_curl = curl_easy_init();

  curl_easy_setopt(p_curl, CURLOPT_WRITEFUNCTION, write_data);
  curl_easy_setopt(p_curl, CURLOPT_WRITEDATA, c_response);

  return p_curl;
}

void* timer_thread(void *p) {
  struct data* p_data = (struct data*) p;
  unsigned long int i;
  time_t t_start_total, t_start;

  t_start_total = t_start = time(NULL);

  for (i = 0;; ++i) {
    time_t t_now, t_elapsed, t_elapsed_total;
    double d_rps, d_perc;

    usleep(1000000);
    t_now = time(NULL);
    t_elapsed_total = t_now - t_start_total;

    if (p_data->i_duration_s <= t_elapsed_total) {
      pthread_mutex_lock(&p_data->mutex);
      p_data->i_exit = 1;
      p_data->i_duration_s = t_elapsed_total;
      pthread_mutex_unlock(&p_data->mutex);

      break;
    }

    t_elapsed = t_now - t_start;

    pthread_mutex_lock(&p_data->mutex);
    d_rps = ((double) p_data->ui_count)/((double) t_elapsed);

    if (p_data->d_rps_wanted > 0) {
      int flag = 0;

      if (d_rps > p_data->d_rps_wanted ) {
        double d_diff = (d_rps - p_data->d_rps_wanted);
        p_data->d_sleep_us = sleep_us(p_data->d_sleep_us, d_rps, p_data->d_rps_wanted, &d_perc);

        fprintf(stdout, "RPS - %d wanted:%d thread_sleep_us:%0.3f sleep_adjust:%0.3f time:%d/%d\n",
            (int) d_rps,
            (int) p_data->d_rps_wanted,
            p_data->d_sleep_us, d_perc,
            (int) t_elapsed_total,
            p_data->i_duration_s);

        flag = 1;

      } else if (d_rps < p_data->d_rps_wanted) {
        double d_diff = (p_data->d_rps_wanted - d_rps);
        p_data->d_sleep_us = sleep_us(p_data->d_sleep_us, d_rps, p_data->d_rps_wanted, &d_perc);

        fprintf(stdout, "RPS + %d wanted:%d thread_sleep_us:%0.3f sleep_adjust:%0.3f time:%d/%d\n",
            (int) d_rps,
            (int) p_data->d_rps_wanted,
            p_data->d_sleep_us, d_perc,
            (int) t_elapsed_total,
            p_data->i_duration_s);

        flag = 1;

      } else {
        fprintf(stdout, "RPS = %d wanted:%d thread_sleep_us:%0.3f time:%d/%d\n",
            (int) d_rps,
            (int) p_data->d_rps_wanted,
            p_data->d_sleep_us,
            (int) t_elapsed_total,
            p_data->i_duration_s);
      }

      if (flag == 1) {
        p_data->ui_count = 0;
        t_start = time(NULL);
      }
    }

    pthread_mutex_unlock(&p_data->mutex);
  }

  return NULL;
}

void* request_thread(void *p) {
  struct data* p_data = (struct data*) p;
  char *c_response = NULL;
  CURL *p_curl = curl_init(&c_response);

  if (p_curl) {
    CURLcode res;
    time_t t1 = time(NULL);
    int i, i_which_url = -1;
    unsigned int d_sleep_us;
    double d_val, d_time_min = 9999999999, d_time_max = -1,
           d_timeout = (double) p_data->i_timeout / (double) 1000,
           d_time_avg = 0;
    unsigned int ui_count = 0, ui_curl_errors = 0, ui_timeout_exceeded = 0, ui_parse_errors = 0;

    for (;; ++ui_count) {
      i_which_url = (i_which_url+1) % p_data->i_num_urls;
      curl_easy_setopt(p_curl, CURLOPT_URL, p_data->pc_urls[i_which_url]);
      res = curl_easy_perform(p_curl);

      if (CURLE_OK == res) {
        if (strlen(c_response) < 1) { 
          fprintf(stderr, "While requesting %s:\n", p_data->pc_urls[i_which_url]);
          fprintf(stderr, "Expected response %s but received empty string\n", p_data->c_expected);
          ++ui_parse_errors;

        } else if (NULL == strstr(c_response, p_data->c_expected)) {
          fprintf(stderr, "While requesting %s:\n", p_data->pc_urls[i_which_url]);
          fprintf(stderr, "Expected response %s but received %s\n", p_data->c_expected, c_response);
          ++ui_parse_errors;
        }

        res = curl_easy_getinfo(p_curl, CURLINFO_TOTAL_TIME, &d_val);

        if ((CURLE_OK == res) && (d_val > 0)) {

          if (d_timeout > 0 && d_val > d_timeout) {
            ++ui_timeout_exceeded;
          }

          if (d_val > d_time_max)
            d_time_max = d_val;

          if (d_val < d_time_min)
            d_time_min = d_val;

          d_time_avg += d_val;
        } else {
          char* c_error = calloc(1024, sizeof(char));
          memset((void*)c_error, 0, 1024);
          c_error[1023] = '\0';
          fprintf(stdout, "(thread %d) cURL returned: %s\n",
              (int) pthread_self(), c_error);
          free(c_error);
          ++ui_curl_errors;
        }
      } else {
        ++ui_curl_errors;
      }

      if (c_response != NULL) {
        free(c_response);
        c_response = NULL;
      }

      if (ui_count % 10000 == 0 && ui_count > 0)
        print_stats(ui_count, time(NULL) - t1, d_time_max, d_time_min, d_time_avg, ui_timeout_exceeded, ui_curl_errors, ui_parse_errors);
      pthread_mutex_lock(&p_data->mutex);
      ++p_data->ui_count;
      d_sleep_us = p_data->d_sleep_us;
      if (p_data->i_exit)
        break;
      pthread_mutex_unlock(&p_data->mutex);


      usleep((unsigned int)d_sleep_us);
    }

    if (p_data->d_time_min > d_time_min)
      p_data->d_time_min = d_time_min;
    if (p_data->d_time_max < d_time_max)
      p_data->d_time_max = d_time_max;

    p_data->ui_curl_errors += ui_curl_errors;
    p_data->ui_parse_errors += ui_parse_errors;
    p_data->ui_timeout_exceeded += ui_timeout_exceeded;
    p_data->ui_count_total += ui_count + 1;
    p_data->d_time_avg += d_time_avg;

    pthread_mutex_unlock(&p_data->mutex);

    curl_easy_cleanup(p_curl);
  }

  return NULL;
}

int is_url(char *c_str) {
  if (strlen(c_str)) {
    if (strncmp("http://", c_str, 7) == 0)
      return 1;
    if (strncmp("https://", c_str, 8) == 0)
      return 1;
  }

  return 0;
}

int main(int i_argc, char **c_argv) {
  struct data d;
  pthread_t *pt_threads;
  pthread_t t_timer;
  int i_threads;
  int i_opt;

  d.c_expected = "";
  d.c_urlparam = NULL;
  d.d_rps_wanted = -1;
  d.i_concurrent = -1;
  d.i_duration_s = -1;
  d.i_timeout = -1;

  while ((i_opt = getopt(i_argc, c_argv, "e:u:r:c:d:t:h")) != -1) {
    switch (i_opt) {
      case 'e':
        d.c_expected = optarg;
        break;
      case 'u':
        d.c_urlparam = optarg;
        break;
      case 'r':
        d.d_rps_wanted = (double) atoi(optarg);
        break;
      case 'c':
        d.i_concurrent = atoi(optarg);
        break;
      case 'd':
        d.i_duration_s = atoi(optarg);
        break;
      case 't':
        d.i_timeout = atoi(optarg);
        break;
      case 'h':
        synopsis();
        exit(SUCCESS);
        break;
      case '?':
        if (optopt == 'e' ||
            optopt == 'u' ||
            optopt == 'r' ||
            optopt == 'c' ||
            optopt == 'd' ||
            optopt == 't'
           )

          fprintf(stderr, "Option -%c requires an argument.\n", optopt);
        else if (isprint(optopt))
          fprintf(stderr, "Unknown option -%c'.\n", optopt);

        else
          fprintf(stderr, "Unknown option character '\\x%x'.\n", optopt);

        exit(E_OPT_ERROR);
      default:
        synopsis();
        exit(E_WRONG_USAGE);
        break;
    }
  }

  checkarg_c('u', d.c_urlparam);
  checkarg_i('r', d.d_rps_wanted);
  checkarg_i('c', d.i_concurrent);
  checkarg_i('d', d.i_duration_s);

  d.i_exit = 0;
  d.d_time_min = 999999;
  d.d_time_max = 0;
  d.ui_timeout_exceeded = 0;
  d.ui_curl_errors = 0;
  d.ui_parse_errors = 0;
  d.ui_count = 0;
  d.ui_count_total = 0;
  d.d_sleep_us = 1000000;
  d.i_num_urls = 0;

  d.pc_urls = calloc(1, sizeof(char*));

  pthread_mutex_init(&d.mutex, NULL);

  if (is_url(d.c_urlparam) == 1) {
    int i_len = strlen(d.c_urlparam) + 1;
    d.pc_urls[d.i_num_urls] = calloc(i_len, sizeof(char));
    strncpy(d.pc_urls[d.i_num_urls], d.c_urlparam, i_len-1);
    d.pc_urls[d.i_num_urls][i_len-1] = '\0';
    ++d.i_num_urls;

  } else {
    // Open urlparamfile
    char c_buff[4096];
    FILE *fh = fopen(d.c_urlparam, "r");

    if (fh == NULL) {
      fprintf(stdout, "Could not open file %s\n", d.c_urlparam);
      exit(E_OPEN_FILE);
    }

    d.i_num_urls = 0;
    while (fgets(c_buff,sizeof(c_buff),fh) != NULL) {
      int i_len = strlen(c_buff);
      d.pc_urls[d.i_num_urls] = calloc(i_len, sizeof(char));
      strncpy(d.pc_urls[d.i_num_urls], c_buff, i_len-1);
      d.pc_urls[d.i_num_urls][i_len-1] = '\0';

      ++d.i_num_urls;
      d.pc_urls = realloc(d.pc_urls, (d.i_num_urls + 1) * sizeof(char*));
    }

    fclose(fh);
  }

  pt_threads = (pthread_t*) calloc(d.i_concurrent, sizeof(pthread_t));

  pthread_create(&t_timer, NULL, timer_thread, (void*) &d);

  for (i_threads = 0; i_threads < d.i_concurrent; ++i_threads) {
    // Add a sleep so threads are mixed up from the beginning
    usleep((unsigned int) i_threads * 10 + 1);
    fprintf(stdout, "Starting thread %d\n", i_threads);
    pthread_create(&pt_threads[i_threads], NULL, request_thread, (void*) &d);
  }

  for (i_threads = 0; i_threads < d.i_concurrent; ++i_threads)
    pthread_join(pt_threads[i_threads], NULL);

  fprintf(stdout, "\nTOTAL STATS using duration:%d concurrent:%d timeout:%d urlparam:%s\n",
      d.i_duration_s,
      d.i_concurrent,
      d.i_timeout,
      d.c_urlparam);
  print_stats(d.ui_count_total,
      d.i_duration_s,
      d.d_time_max,
      d.d_time_min,
      d.d_time_avg,
      d.ui_timeout_exceeded,
      d.ui_curl_errors,
      d.ui_parse_errors);

  // Cleaning up stuff
  for (; 0 < d.i_num_urls; --d.i_num_urls)
    free(d.pc_urls[d.i_num_urls-1]);

  free(d.pc_urls);

  pthread_mutex_destroy(&d.mutex);

  return SUCCESS;
}
